### PR TITLE
Implement ZIO#withConfigProvider

### LINKED ISSUE #7558: ZIO 2.0 Integrated Config improvements
- [x] Introduce `ZIO#withConfigProvider`, that supplies a config provider for the scope of the specified effect.
 - [x] Ensure the `Chunk` parameters to the `Config.Error` constructors all have defaults of `Chunk.empty`, so users do not have to supply a path prefix themselves.
 - [x] Create a helper `Config#validate(error: => String)(pf: PartialFunction[A, B]): Config[B]` which creates a validation error from the specified `String` if the partial function does not yield a `B`.
 - [x] Create a subtype of `Fallback` called `Optional`, with an `unapply` and `apply`. Then use this subtype to implement `Config#optional`, so that it becomes possible for some `ConfigProvider` to pattern match on optional values and treat them specially.

Review by @vigoo

### PR DESCRIPTION
**Pull request by adamgfraser**
Resolves #7558.

### CODE REVIEW

**Review by jdegoes — DISMISSED** [review]

**Comment by jdegoes** [comment]
Awesome! 🚀 ❤️ 

Did you want to do `Optional` in this or should I put it into a separate ticket?

**Comment by adamgfraser** [comment]
@jdegoes Will do here.

**Review by jdegoes — APPROVED** [review]

**Review by vigoo — APPROVED** [review]

File: core/shared/src/main/scala/zio/Config.scala [hunk 1]
```diff
@@ -187,7 +196,33 @@ object Config {
...
+    def productElement(n: Int): Any =
+      n match {
+        case 0 => first
+        case 1 => second
+        case _ => throw new IndexOutOfBoundsException(n.toString)
+      }
+  }
+  object Fallback {
+    def apply[A](first: Config[A], second: Config[A]): Fallback[A]        = new Fallback(first, second)
+    def unapply[A](fallback: Fallback[A]): Option[(Config[A], Config[A])] = Some((fallback.first, fallback.second))
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/Config.scala:223]: Btw you can type this as `Some` instead of `Option`.

Comment (adamgfraser) [reply core/shared/src/main/scala/zio/Config.scala:223]: I actually did it that way initially but it fails binary compatibility.

Comment (jdegoes) [reply core/shared/src/main/scala/zio/Config.scala:223]: MIMA doesn't understand covariance? 🤔

Comment (adamgfraser) [reply core/shared/src/main/scala/zio/Config.scala:223]: Apparently. We can change it and ignore the MIMA problem but I figured better safe than sorry.

**Pull request merged**